Struct types

flodiebold opened the pull request:
Infer types for struct fields, and add basic field completions. There's also some code for enums, but I focused on getting structs working.

There's still ways to go before this becomes useful: There's no autoderef (or even reference types) and no inference for `self`, for example.

matklad commented on crates/ra_analysis/src/completion.rs:
```diff
@@ -20,13 +21,13 @@ use crate::{
-/// incomplete and can look readlly weired.
+/// incomplete and can look really weird.
 ///
-/// Once the context is collected, we run a series of completion routines whihc
```
One day, I'll bolt on a spell checker on top of rust-analyzer... Correct spell checking  really needs to know the language for the following features:

* distinguishing between definitions and references. Only defs should be spell checked. 
* handling escapes in string literals
* handling doc-syntax of comments

matklad commented on crates/ra_analysis/src/completion/complete_dot.rs:
```diff
@@ -0,0 +1,98 @@
+        return Ok(());
+    };
+    let function = if let Some(fn_def) = ctx.enclosing_fn {
+        hir::source_binder::function_from_module(ctx.db, module, fn_def)
```
Note to self: this should be handled during context creation. We also wrongly use source_binder in `complete_scope`.

matklad commented on crates/ra_analysis/src/completion/complete_dot.rs:
```diff
@@ -0,0 +1,98 @@
+    Ok(())
+}
+
+fn complete_fields(acc: &mut Completions, ctx: &CompletionContext, ty: Ty) -> Cancelable<()> {
```
Perhaps

```suggestion
fn complete_fields(acc: &mut Completions, ctx: &CompletionContext, receiver: Ty) -> Cancelable<()> {
```

?

matklad commented on crates/ra_analysis/src/completion/completion_context.rs:
```diff
@@ -145,6 +152,23 @@ impl<'a> CompletionContext<'a> {
+                .ancestors()
+                .take_while(|it| it.kind() != SOURCE_FILE && it.kind() != MODULE)
+                .find_map(ast::FieldExpr::cast)
+                .and_then(ast::FieldExpr::expr);
```
I feel that this could break, because leaf can be an arbitrary thing:

```
foo
  /* comment */
  .
  /* more comments */
  cursor_here
```

We need to figure out a more robust pattern here. In intelliJ, this  "patched file"

```
        let file = {
            let edit = AtomTextEdit::insert(offset, "intellijRulezz".to_string());
            original_file.reparse(&edit)
        };
```

is a long-lived thing and all completion infra uses it. However, this leads to a ton of bugs when you use patched file where the original file is expected. That's why in ra I try to strictly separate the "gather context" and "compute completion" bits.


I think what would work here is to find the receiver expr in the "patched" file, get its range, and then find an expression with this range in the original file.

flodiebold replied:
Ok, done :white_check_mark:

matklad commented on crates/ra_hir/src/adt.rs:
```diff
@@ -0,0 +1,213 @@
+        let name = struct_def
+            .name()
+            .map(|n| n.text())
+            .unwrap_or(SmolStr::new("[error]"));
```
Interesting! It seems easier to just not have anonymous items in the hir. But then you'll be unable to get the type of the field within the struct itself. So perhaps having a sort-of NaN name would work better?

matklad replied:
OTOH, we don't need  "to get the type of the field within the struct itself" so let's perhaps just return `None` if the name is missing?

flodiebold replied:
Hm. I think I'll make the name an option. We might still want to get the HIR for a struct without name (though it's admittedly unlikely).

matklad commented on crates/ra_hir/src/adt.rs:
```diff
@@ -0,0 +1,213 @@
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StructField {
+    name: SmolStr,
+    ty: Ty,
```
This will eagerly compute the `Ty` of the field. I think it's not a big deal, but perhaps it's better long-term if by default everything is lazy? That is, getting the set of names of fields is one query, getting the type of a specific field is another query?

flodiebold replied:
Yeah, I was considering introducing something like `hir::Ty` (we'll need it when introducing `hir::Expr` at the latest). I was then wondering whether that should already be resolved; rustc's HIR already contains the resolved defs, but the resolution is the only potentially expensive thing about constructing a `ty::Ty` at the moment. So I thought I'd leave it for now.

matklad commented on crates/ra_hir/src/db.rs:
```diff
@@ -31,6 +32,16 @@ pub trait HirDatabase: SyntaxDatabase
+        use fn query_definitions::struct_data;
+    }
+
+    fn enum_data(def_id: DefId) -> Cancelable<Arc<EnumData>> {
```
Note that we have a newtyped `FnId` for functions. I think we should either use type-safe def-ids everywhere, or a single DefId everywhere.

matklad replied:
Makes sense, yeah. I think I'll remove `FnId`, and will make `Module` a newtype of `DefId` as well.

flodiebold commented:
@matklad I've added a `hir::TypeRef` as an intermediate between `ast::TypeRef` and `ty::Ty`, and used that to resolve the field types lazily.

I also added reference / pointer types since I was already working on them and it intersected a bit :smile:

matklad commented:
bors r+

Super!

matklad commented:
>@matklad I've added a hir::TypeRef as an intermediate between ast::TypeRef and ty::Ty, and used that to resolve the field types lazily.

One thing to think about is how would we map *back* from `hit::TypeRef` to `ast::TypeRef` (this will be needed in IDE bits I think, and for error reporting). An obvious solution would be to store spans/syntax ptrs inside the hir, but I'd rather avoid that: this information takes space, and is only rarely needed. Ideally, we invent some kind of side-table approach here....
